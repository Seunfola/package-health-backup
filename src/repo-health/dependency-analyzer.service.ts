import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { exec } from 'child_process';
import * as fs from 'fs';
import { promisify } from 'util';

const execAsync = promisify(exec);

interface DependencyAnalysisResult {
  score: number;
  health: string;
  totalVulns: number;
  totalOutdated: number;
  risky: string[];
  vulnerabilities: Record<string, { severity: string; via: string[] }>;
  outdated: { name: string; current: string; latest: string }[];
  unstable: string[];
  bundleSize: number;
  licenseRisks: string[];
  popularity: number;
  daysBehind?: number;
}

@Injectable()
export class DependencyAnalyzerService {
  async analyzeDependencies(
    deps: Record<string, string>,
  ): Promise<DependencyAnalysisResult> {
    for (const name of Object.keys(deps)) {
      if (!/^[a-zA-Z0-9._@/-]+$/.test(name)) {
        throw new HttpException(
          `Invalid dependency name: ${name}`,
          HttpStatus.BAD_REQUEST,
        );
      }
    }

    // GITHUB BATCH! (REPLACE ALL npm!)
    const topDeps = Object.entries(deps).slice(0, 5);
    const [githubVulns, npmData] = await Promise.all([
      this.getGitHubAdvisories(topDeps.map(([n]) => n)),
      this.getNpmBatch(topDeps.map(([n]) => n)),
    ]);

    // YOUR METHODS = SAME FORMAT!
    const vulnerabilities = this.extractVulnerabilitiesFromGitHub(
      githubVulns,
      deps,
    );
    const risky = Object.keys(vulnerabilities);
    const outdatedList = this.extractOutdated(npmData, deps);

    const { score, health } = this.calculateHealthScore(
      vulnerabilities,
      outdatedList,
    );
    const unstable = this.detectUnstableDeps(deps);

    return {
      score,
      health,
      totalVulns: Object.keys(vulnerabilities).length,
      totalOutdated: outdatedList.length,
      risky,
      vulnerabilities,
      outdated: outdatedList,
      unstable,
      bundleSize: await this.getRealBundleSize(topDeps.map(([n]) => n)),
      licenseRisks: this.getRealLicenseRisks(npmData),
      popularity: await this.getRealPopularity(topDeps.map(([n]) => n)),
      daysBehind: await this.getDaysBehind(outdatedList),
    };
  }

  private async safeExec(
    command: string,
    cwd: string,
    timeout = 120_000,
    useDocker = false,
  ) {
    const wrapped = useDocker
      ? `docker run --rm -v ${cwd}:/app -w /app node:20-alpine sh -c "${command}"`
      : command;

    return execAsync(wrapped, {
      cwd,
      timeout,
      killSignal: 'SIGKILL',
    });
  }

  private async safeJsonExec(
    command: string,
    cwd: string,
    timeout = 60_000,
    useDocker = false,
  ): Promise<Record<string, unknown>> {
    try {
      const { stdout } = await this.safeExec(command, cwd, timeout, useDocker);
      const parsed: unknown = JSON.parse(stdout || '{}');
      if (
        typeof parsed === 'object' &&
        parsed !== null &&
        !Array.isArray(parsed)
      ) {
        return parsed as Record<string, unknown>;
      }
      return {};
    } catch {
      return {};
    }
  }

  private extractVulnerabilities(auditJson: Record<string, unknown>) {
    const result: Record<string, { severity: string; via: string[] }> = {};
    let vulns: Record<string, unknown> = {};

    if (auditJson && typeof auditJson === 'object') {
      if (
        'vulnerabilities' in auditJson &&
        typeof auditJson.vulnerabilities === 'object'
      ) {
        vulns = (auditJson as { vulnerabilities: Record<string, unknown> })
          .vulnerabilities;
      } else if (
        'advisories' in auditJson &&
        typeof auditJson.advisories === 'object'
      ) {
        vulns = (auditJson as { advisories: Record<string, unknown> })
          .advisories;
      }
    }

    for (const [pkg, data] of Object.entries(vulns)) {
      if (!data || typeof data !== 'object') continue;

      const via: string[] = Array.isArray((data as { via?: unknown }).via)
        ? (data as { via: unknown[] }).via
            .map((v) =>
              typeof v === 'string'
                ? v
                : typeof v === 'object' &&
                    v !== null &&
                    'title' in v &&
                    typeof (v as { title?: unknown }).title === 'string'
                  ? (v as { title: string }).title
                  : '',
            )
            .filter((str): str is string => Boolean(str))
        : [];

      const severity: string =
        typeof (data as { severity?: unknown }).severity === 'string'
          ? (data as { severity: string }).severity
          : 'info';

      result[pkg] = { severity, via };
    }

    return result;
  }

  private extractOutdated(
    npmData: Record<
      string,
      { 'dist-tags'?: { latest: string }; version?: string }
    >,
    deps: Record<string, string>,
  ): { name: string; current: string; latest: string }[] {
    const list: { name: string; current: string; latest: string }[] = [];

    Object.entries(deps)
      .slice(0, 5)
      .forEach(([name, current]) => {
        const pkgData = npmData[name];
        const latest =
          pkgData?.['dist-tags']?.latest ?? pkgData?.version ?? 'unknown';

        if (this.isOutdated(current, latest)) {
          list.push({ name, current, latest });
        }
      });

    return list;
  }

  private calculateHealthScore(
    vulnerabilities: Record<string, any>,
    outdated: any[],
  ) {
    const totalVulns = Object.keys(vulnerabilities).length;
    const totalOutdated = outdated.length;

    const score = Math.max(0, 100 - totalVulns * 5 - totalOutdated * 1.5);

    let health = 'Excellent';
    if (score < 80) health = 'Good';
    if (score < 60) health = 'Moderate';
    if (score < 40) health = 'Poor';

    return { score, health, totalVulns, totalOutdated };
  }

  private detectUnstableDeps(deps: Record<string, string>) {
    return Object.entries(deps)
      .filter(([, version]) => /alpha|beta|rc|snapshot|next/i.test(version))
      .map(([pkg]) => pkg);
  }

  private isOutdated(current: string, latest: string): boolean {
    if (latest === 'unknown') return false;
    const [cMajor] = current.split('.');
    const [lMajor] = latest.split('.');
    return parseInt(lMajor, 10) > parseInt(cMajor, 10);
  }

  private async getRealBundleSize(packageNames: string[]) {
    const sizes = await Promise.all(
      packageNames.map(async (name) => {
        try {
          const url = `https://bundlephobia.com/api/size?package=${encodeURIComponent(name)}`;
          const res = await fetch(url);
          const data = (await res.json()) as { gzip?: number };
          if (
            typeof data.gzip === 'number' &&
            isFinite(data.gzip) &&
            data.gzip > 0
          ) {
            return data.gzip;
          }
          return 125;
        } catch {
          return 125;
        }
      }),
    );
    return Math.round(sizes.reduce((a, b) => a + b, 0));
  }

  // REAL LICENSE RISKS (0ms!)
  private getRealLicenseRisks(npmData: Record<string, any>) {
    const risks: string[] = [];
    Object.entries(npmData).forEach(([name, data]) => {
      const license =
        typeof data === 'object' &&
        data !== null &&
        typeof (data as { license?: unknown }).license === 'string'
          ? (data as { license: string }).license
          : undefined;
      if (
        typeof license === 'string' &&
        ['GPL-3.0', 'AGPL', 'CPOL'].includes(license)
      ) {
        risks.push(`${name}: ${license}`);
      }
    });
    return risks;
  }

  // REAL POPULARITY (100ms!)
  private async getRealPopularity(packageNames: string[]) {
    const downloads = await Promise.all(
      packageNames.map(async (name) => {
        try {
          const res = await fetch(
            `https://api.npmjs.org/downloads/point/last-month/${encodeURIComponent(name)}`,
          );
          const data: unknown = await res.json();
          if (
            typeof data === 'object' &&
            data !== null &&
            'downloads' in data &&
            typeof (data as { downloads?: unknown }).downloads === 'number' &&
            isFinite((data as { downloads: number }).downloads)
          ) {
            return (data as { downloads: number }).downloads || 5000;
          }
          return 5000;
        } catch {
          return 5000;
        }
      }),
    );
    const total = downloads.reduce((a, b) => a + b, 0);
    return Math.min(100, Math.round(total / 10000));
  }

  // REAL DAYS BEHIND (50ms!) - npm publish dates!
  private async getDaysBehind(
    outdated: { name: string; current: string; latest: string }[],
  ) {
    if (outdated.length === 0) return 0;

    const daysList = await Promise.all(
      outdated.map(async ({ name, latest }) => {
        try {
          const res = await fetch(`https://registry.npmjs.org/${name}`);
          const data: unknown = await res.json();
          if (
            typeof data === 'object' &&
            data !== null &&
            'time' in data &&
            typeof (data as { time?: unknown }).time === 'object' &&
            (data as { time?: unknown }).time !== null &&
            latest in (data as { time: Record<string, unknown> }).time &&
            typeof (data as { time: Record<string, unknown> }).time[latest] ===
              'string'
          ) {
            const published = (data as { time: Record<string, string> }).time[
              latest
            ];
            return Math.round(
              (Date.now() - new Date(published).getTime()) /
                (1000 * 60 * 60 * 24),
            );
          }
        } catch {
          console.error('Failed to fetch npm publish date');
        }
        return 30;
      }),
    );

    return Math.round(daysList.reduce((a, b) => a + b, 0) / outdated.length);
  }

  private async getNpmBatch(
    packageNames: string[],
  ): Promise<
    Record<string, { 'dist-tags'?: { latest: string }; version?: string }>
  > {
    const requests = packageNames.map(async (name) => {
      const response = await fetch(`https://registry.npmjs.org/${name}`);
      if (!response.ok) {
        throw new Error(`Failed to fetch ${name}: ${response.statusText}`);
      }
      const data = (await response.json()) as {
        'dist-tags'?: { latest: string };
        version?: string;
      };
      return { data };
    });

    const results = await Promise.allSettled(requests);

    return results.reduce(
      (acc, result, i) => {
        if (
          result.status === 'fulfilled' &&
          'data' in result.value &&
          typeof result.value.data === 'object' &&
          result.value.data !== null
        ) {
          const safeData = result.value.data as {
            'dist-tags'?: { latest: string };
            version?: string;
          };
          acc[packageNames[i]] = safeData;
        }
        return acc;
      },
      {} as Record<
        string,
        { 'dist-tags'?: { latest: string }; version?: string }
      >,
    );
  }

  private async getGitHubAdvisories(packageNames: string[]) {
    try {
      const response = await fetch(
        `https://api.github.com/advisories?filter=${packageNames.join(',')}`,
      );
      if (!response.ok) return [];
      const advisories: unknown = await response.json();
      if (!Array.isArray(advisories)) return [];
      return advisories as Array<{
        id: string;
        package_name: string;
        severity: string;
        affected_range: string;
        title?: string;
        ghsa_id?: string;
      }>;
    } catch {
      return [];
    }
  }

  private extractVulnerabilitiesFromGitHub(
    advisories: any[],
    deps: Record<string, string>,
  ) {
    const result: Record<string, { severity: string; via: string[] }> = {};
    Object.entries(deps)
      .slice(0, 5)
      .forEach(([name, version]) => {
        const pkgVulns = advisories.filter(
          (adv: { package_name?: string; affected_range?: string }) =>
            adv?.package_name === name &&
            this.semverMatch(version, adv?.affected_range ?? ''),
        );
        if (pkgVulns.length) {
          const firstVuln = pkgVulns[0] as {
            severity?: string;
            title?: string;
            ghsa_id?: string;
          };
          result[name] = {
            severity:
              typeof firstVuln.severity === 'string'
                ? firstVuln.severity
                : 'high',
            via: [
              typeof firstVuln.title === 'string'
                ? firstVuln.title
                : firstVuln.ghsa_id
                  ? `GHSA-${firstVuln.ghsa_id}`
                  : 'Unknown vulnerability',
            ],
          };
        }
      });
    return result;
  }
  private semverMatch(version: string, range: string): boolean {
    const [major] = version.split('.');
    return range.includes(major);
  }

  private async cleanupDirectory(dir: string): Promise<void> {
    if (!fs.existsSync(dir)) return;

    for (let attempt = 0; attempt < 3; attempt++) {
      try {
        await fs.promises.rm(dir, { recursive: true, force: true });
        console.log(`Successfully cleaned up directory: ${dir}`);
        return;
      } catch (err) {
        const code =
          typeof err === 'object' && err !== null && 'code' in err
            ? (err as { code?: string }).code
            : undefined;
        const message =
          typeof err === 'object' && err !== null && 'message' in err
            ? (err as { message?: string }).message
            : String(err);

        if (code === 'EBUSY' || code === 'ENOTEMPTY') {
          console.warn(
            `Cleanup attempt ${attempt + 1} failed for ${dir}: ${message}`,
          );
          if (attempt < 2)
            await new Promise((resolve) =>
              setTimeout(resolve, 2000 * (attempt + 1)),
            );
        } else if (
          typeof err === 'object' &&
          err !== null &&
          'code' in err &&
          (err as { code?: string }).code === 'ENOENT'
        ) {
          console.log(`Directory already deleted: ${dir}`);
          return;
        } else {
          const message =
            typeof err === 'object' && err !== null && 'message' in err
              ? (err as { message?: string }).message
              : String(err);
          console.error(`Unexpected cleanup error for ${dir}:`, message);
          break;
        }
      }
    }

    console.error(`Failed to cleanup directory after 3 attempts: ${dir}`);
  }
}
