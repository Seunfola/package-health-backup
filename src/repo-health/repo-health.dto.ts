import { ApiProperty } from '@nestjs/swagger';

export class AnalyzePublicRepoDto {
  @ApiProperty({ example: 'nestjs', description: 'Repository owner' })
  owner!: string;

  @ApiProperty({ example: 'nest', description: 'Repository name' })
  repo!: string;

  @ApiProperty({ required: false, description: 'Package.json content' })
  packageJson?: string;
}

export class AnalyzePrivateRepoDto {
  @ApiProperty({ example: 'your-username', description: 'Repository owner' })
  owner!: string;

  @ApiProperty({ example: 'your-private-repo', description: 'Repository name' })
  repo!: string;

  @ApiProperty({
    example: 'ghp_xxx',
    description: 'GitHub token (required for private repos)',
  })
  token!: string;

  @ApiProperty({ required: false, description: 'Package.json content' })
  packageJson?: string;
}

export class AnalyzePublicUrlDto {
  @ApiProperty({
    example: 'https://github.com/nestjs/nest',
    description: 'GitHub repository URL',
  })
  url!: string;

  @ApiProperty({ required: false, description: 'Package.json content' })
  packageJson?: string;
}

export class AnalyzePrivateUrlDto {
  @ApiProperty({
    example: 'https://github.com/username/private-repo',
    description: 'GitHub repository URL',
  })
  url!: string;

  @ApiProperty({
    example: 'ghp_xxx',
    description: 'GitHub token (required for private repos)',
  })
  token!: string;

  @ApiProperty({ required: false, description: 'Package.json content' })
  packageJson?: string;
}

export class AnalyzeAutoRepoDto {
  @ApiProperty({ example: 'nestjs', description: 'Repository owner' })
  owner!: string;

  @ApiProperty({ example: 'nest', description: 'Repository name' })
  repo!: string;

  @ApiProperty({
    required: false,
    example: 'ghp_xxx',
    description: 'GitHub token (required only if repository is private)',
  })
  token?: string;

  @ApiProperty({ required: false, description: 'Package.json content' })
  packageJson?: string;
}

export class AnalyzeAutoUrlDto {
  @ApiProperty({
    example: 'https://github.com/nestjs/nest',
    description: 'GitHub repository URL',
  })
  url!: string;

  @ApiProperty({
    required: false,
    example: 'ghp_xxx',
    description: 'GitHub token (required only if repository is private)',
  })
  token?: string;

  @ApiProperty({ required: false, description: 'Package.json content' })
  packageJson?: string;
}