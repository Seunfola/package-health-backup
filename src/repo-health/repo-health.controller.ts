// repo-health.controller.ts
import {
  Controller,
  Post,
  Body,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiBody } from '@nestjs/swagger';
import { RepoHealthService } from './services/repo-health.service';
import {
  AnalyzePublicRepoDto,
  AnalyzePrivateRepoDto,
  AnalyzePublicUrlDto,
  AnalyzePrivateUrlDto,
  AnalyzeAutoRepoDto,
  AnalyzeAutoUrlDto,
} from './repo-health.dto';

@ApiTags('repo-health')
@Controller('repo-health')
export class RepoHealthController {
  constructor(private readonly repoHealthService: RepoHealthService) {}

  @Post('public/repo')
  @ApiOperation({
    summary: 'Analyze PUBLIC repository by owner/repo',
    description: 'No token required for public repositories',
  })
  @ApiBody({ type: AnalyzePublicRepoDto })
  async analyzePublicRepo(@Body() body: AnalyzePublicRepoDto) {
    const { owner, repo, packageJson } = body;

    if (!owner || !repo) {
      throw new HttpException(
        'Both owner and repo are required',
        HttpStatus.BAD_REQUEST,
      );
    }

    try {
      const rawJson = packageJson ? JSON.parse(packageJson) : undefined;
      return await this.repoHealthService.analyzePublicRepository(
        owner,
        repo,
        undefined, // file
        rawJson,
      );
    } catch (err: unknown) {
      if (err instanceof HttpException) throw err;
      const message =
        err instanceof Error ? err.message : 'Unexpected error occurred';
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  @Post('public/url')
  @ApiOperation({
    summary: 'Analyze PUBLIC repository by URL',
    description: 'No token required for public repositories',
  })
  @ApiBody({ type: AnalyzePublicUrlDto })
  async analyzePublicRepoByUrl(@Body() body: AnalyzePublicUrlDto) {
    const { url, packageJson } = body;

    if (!url) {
      throw new HttpException('GitHub URL is required', HttpStatus.BAD_REQUEST);
    }

    try {
      const rawJson = packageJson ? JSON.parse(packageJson) : undefined;
      return await this.repoHealthService.analyzePublicRepoByUrl(
        url,
        undefined, // file
        rawJson,
      );
    } catch (err: unknown) {
      if (err instanceof HttpException) throw err;
      const message =
        err instanceof Error ? err.message : 'Unexpected error occurred';
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  @Post('private/repo')
  @ApiOperation({
    summary: 'Analyze PRIVATE repository by owner/repo',
    description: 'Token is REQUIRED for private repositories',
  })
  @ApiBody({ type: AnalyzePrivateRepoDto })
  async analyzePrivateRepo(@Body() body: AnalyzePrivateRepoDto) {
    const { owner, repo, token, packageJson } = body;

    if (!owner || !repo || !token) {
      throw new HttpException(
        'Owner, repo, and token are required for private repositories',
        HttpStatus.BAD_REQUEST,
      );
    }

    try {
      const rawJson = packageJson ? JSON.parse(packageJson) : undefined;
      return await this.repoHealthService.analyzePrivateRepository(
        owner,
        repo,
        token, // required
        undefined, // file
        rawJson,
      );
    } catch (err: unknown) {
      if (err instanceof HttpException) throw err;
      const message =
        err instanceof Error ? err.message : 'Unexpected error occurred';
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  @Post('private/url')
  @ApiOperation({
    summary: 'Analyze PRIVATE repository by URL',
    description: 'Token is REQUIRED for private repositories',
  })
  @ApiBody({ type: AnalyzePrivateUrlDto })
  async analyzePrivateRepoByUrl(@Body() body: AnalyzePrivateUrlDto) {
    const { url, token, packageJson } = body;

    if (!url || !token) {
      throw new HttpException(
        'URL and token are required for private repositories',
        HttpStatus.BAD_REQUEST,
      );
    }

    try {
      const rawJson = packageJson ? JSON.parse(packageJson) : undefined;
      return await this.repoHealthService.analyzePrivateRepoByUrl(
        url,
        token, // required
        undefined, // file
        rawJson,
      );
    } catch (err: unknown) {
      if (err instanceof HttpException) throw err;
      const message =
        err instanceof Error ? err.message : 'Unexpected error occurred';
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  @Post('analyze/auto')
  @ApiOperation({
    summary: 'Auto-detect repository type and analyze',
    description: 'Token is required only if repository is private',
  })
  @ApiBody({
    type: AnalyzeAutoRepoDto,
    description: 'Auto-detect repository visibility and analyze accordingly',
  })
  async analyzeRepoAuto(@Body() body: AnalyzeAutoRepoDto) {
    const { owner, repo, token, packageJson } = body;

    if (!owner || !repo) {
      throw new HttpException(
        'Both owner and repo are required',
        HttpStatus.BAD_REQUEST,
      );
    }

    try {
      const rawJson = packageJson ? JSON.parse(packageJson) : undefined;
      return await this.repoHealthService.analyzeRepositoryAuto(
        owner,
        repo,
        undefined, // file
        rawJson,
        token,
      );
    } catch (err: unknown) {
      if (err instanceof HttpException) throw err;
      const message =
        err instanceof Error ? err.message : 'Unexpected error occurred';
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  @Post('analyze/url-auto')
  @ApiOperation({
    summary: 'Auto-detect repository type by URL and analyze',
    description: 'Token is required only if repository is private',
  })
  @ApiBody({
    type: AnalyzeAutoUrlDto,
    description:
      'Auto-detect repository visibility from URL and analyze accordingly',
  })
  async analyzeByUrlAuto(@Body() body: AnalyzeAutoUrlDto) {
    const { url, token, packageJson } = body;

    if (!url) {
      throw new HttpException('GitHub URL is required', HttpStatus.BAD_REQUEST);
    }

    try {
      const rawJson = packageJson ? JSON.parse(packageJson) : undefined;
      return await this.repoHealthService.analyzeByUrlAuto(
        url,
        undefined, // file
        rawJson,
        token,
      );
    } catch (err: unknown) {
      if (err instanceof HttpException) throw err;
      const message =
        err instanceof Error ? err.message : 'Unexpected error occurred';
      throw new HttpException(message, HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }


  @Post('analyze')
  @ApiOperation({
    summary: 'Analyze repository (legacy endpoint)',
    description:
      'Legacy endpoint for backward compatibility. Uses auto-detection.',
  })
  @ApiBody({ type: AnalyzeAutoRepoDto })
  async analyzeRepoLegacy(@Body() body: AnalyzeAutoRepoDto) {
    return this.analyzeRepoAuto(body);
  }

  @Post('analyze-url')
  @ApiOperation({
    summary: 'Analyze repository by URL (legacy endpoint)',
    description:
      'Legacy endpoint for backward compatibility. Uses auto-detection.',
  })
  @ApiBody({ type: AnalyzeAutoUrlDto })
  async analyzeByUrlLegacy(@Body() body: AnalyzeAutoUrlDto) {
    return this.analyzeByUrlAuto(body);
  }
}
