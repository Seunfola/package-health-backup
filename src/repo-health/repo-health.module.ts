import { Module } from '@nestjs/common';
import { RepoHealthService } from './repo-health/repo-health.service';
import { RepoHealthController } from './repo-health/repo-health.controller';

@Module({
  providers: [RepoHealthService],
  controllers: [RepoHealthController]
})
export class RepoHealthModule {}
