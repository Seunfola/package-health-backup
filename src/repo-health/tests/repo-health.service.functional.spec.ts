// repo-health.service.functional.spec.ts
import { Test, TestingModule } from '@nestjs/testing';
import { RepoHealthService } from '../services/repo-health.service';
import { RepositoryDataService } from '../services/repository-data.service';
import { GithubApiService } from '../services/github-api.service';
import { DependencyAnalysisService } from '../services/dependency-analysis.service';
import { HealthCalculatorService } from '../services/health-calculator.service';


// Enhanced mock services that handle the complete schema
class MockGithubApiService {
  async fetchPublicRepositoryData(owner: string, repo: string) {
    return {
      name: `${repo}`,
      stargazers_count: Math.floor(Math.random() * 1000) + 100,
      forks_count: Math.floor(Math.random() * 500) + 50,
      open_issues_count: Math.floor(Math.random() * 50) + 5,
      pushed_at: new Date().toISOString(),
    };
  }

  async fetchPublicCommitActivity(owner: string, repo: string) {
    return Array.from({ length: 4 }, (_, i) => ({
      total: Math.floor(Math.random() * 100) + 10,
    }));
  }

  async fetchPublicSecurityAlerts(owner: string, repo: string) {
    return repo.includes('vulnerable')
      ? [{ severity: 'high' }, { severity: 'medium' }]
      : [];
  }

  async determineRepoVisibility(owner: string, repo: string, token?: string) {
    return token ? 'private' : 'public';
  }
}

class MockDependencyAnalysisService {
  async analyzeDependencies(
    file?: Express.Multer.File,
    rawJson?: string | Record<string, unknown>,
  ) {
    const hasRisks = file || rawJson;

    return {
      dependencyHealth: { score: hasRisks ? 75 : 90 },
      riskyDependencies: hasRisks ? ['vulnerable-pkg@1.0.0'] : [],
      bundleSize: 2048,
      licenseRisks: hasRisks ? ['GPL-3.0'] : ['MIT'],
      popularity: hasRisks ? 80 : 95,
      daysBehind: hasRisks ? 15 : 2,
    };
  }
}

class MockHealthCalculatorService {
  calculateHealthScore(
    repoData: any,
    commitActivity: any[],
    securityAlerts: any[],
    dependencyHealth: number,
  ) {
    const activityScore =
      commitActivity.reduce((sum, item) => sum + item.total, 0) /
      commitActivity.length;
    const securityScore =
      securityAlerts.length === 0
        ? 100
        : Math.max(0, 100 - securityAlerts.length * 10);

    return {
      score: Math.round((activityScore + securityScore + dependencyHealth) / 3),
      metrics: {
        activity: Math.round(activityScore),
        security: securityScore,
        maintenance: dependencyHealth,
        popularity: repoData.stargazers_count > 500 ? 95 : 75,
      },
    };
  }
}

class MockRepositoryDataService {
  private mockData = new Map();

  async findOne(repoId: string) {
    return this.mockData.get(repoId) || null;
  }

  async upsertRepoHealth(repoId: string, data: any) {
    const record = {
      ...data,
      _id: `mock_${repoId.replace('/', '_')}`,
      createdAt: new Date(),
      updatedAt: new Date(),
    };
    this.mockData.set(repoId, record);
    return record;
  }

  async findAll(page = 1, limit = 10, sort = 'stars') {
    const data = Array.from(this.mockData.values());
    return {
      data: data.slice((page - 1) * limit, page * limit),
      total: data.length,
      page,
      limit,
      totalPages: Math.ceil(data.length / limit),
    };
  }

  async findByOwner(owner: string) {
    return Array.from(this.mockData.values()).filter(
      (repo: any) => repo.owner === owner,
    );
  }

  async getStats() {
    const data = Array.from(this.mockData.values());
    const scores = data.map((repo: any) => repo.overall_health.score);
    const average = scores.reduce((a, b) => a + b, 0) / scores.length;

    return {
      totalRepos: data.length,
      averageHealth: Math.round(average * 100) / 100,
      healthDistribution: {
        excellent: data.filter((repo: any) => repo.overall_health.score >= 90)
          .length,
        good: data.filter(
          (repo: any) =>
            repo.overall_health.score >= 70 && repo.overall_health.score < 90,
        ).length,
        fair: data.filter(
          (repo: any) =>
            repo.overall_health.score >= 50 && repo.overall_health.score < 70,
        ).length,
        poor: data.filter((repo: any) => repo.overall_health.score < 50).length,
      },
    };
  }
}

describe('RepoHealthService Functional', () => {
  let service: RepoHealthService;
  let repositoryDataService: RepositoryDataService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        RepoHealthService,
        {
          provide: GithubApiService,
          useClass: MockGithubApiService,
        },
        {
          provide: DependencyAnalysisService,
          useClass: MockDependencyAnalysisService,
        },
        {
          provide: HealthCalculatorService,
          useClass: MockHealthCalculatorService,
        },
        {
          provide: RepositoryDataService,
          useClass: MockRepositoryDataService,
        },
      ],
    }).compile();

    service = module.get<RepoHealthService>(RepoHealthService);
    repositoryDataService = module.get<RepositoryDataService>(
      RepositoryDataService,
    );
  });

  describe('Complete Repository Analysis Lifecycle', () => {
    it('should perform end-to-end analysis and storage with all schema fields', async () => {
      // Execute
      const result = await service.analyzePublicRepository('facebook', 'react');

      // Assert - Verify all schema fields are populated
      expect(result).toMatchObject({
        repo_id: 'facebook/react',
        owner: 'facebook',
        repo: 'react',
        name: 'react',
        stars: expect.any(Number),
        forks: expect.any(Number),
        open_issues: expect.any(Number),
        last_pushed: expect.any(Date),
        commit_activity: expect.any(Array),
        security_alerts: expect.any(Number),
        dependency_health: expect.any(Number),
        risky_dependencies: expect.any(Array),
        overall_health: {
          score: expect.any(Number),
          metrics: expect.any(Object),
        },
        bundle_size: 2048,
        license_risks: expect.any(Array),
        popularity: expect.any(Number),
        days_behind: expect.any(Number),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });

      // Verify the record can be retrieved
      const retrieved = await service.findRepoHealth('facebook', 'react');
      expect(retrieved.repo_id).toBe('facebook/react');
      expect(retrieved.overall_health.score).toBeGreaterThan(0);
    });

    it('should handle repositories with security vulnerabilities', async () => {
      // Execute
      const result = await service.analyzePublicRepository(
        'owner',
        'vulnerable-repo',
      );

      // Assert
      expect(result.security_alerts).toBeGreaterThan(0);
      expect((result.overall_health as any).metrics.security).toBeLessThan(100);
    });

    it('should handle dependency analysis with risky packages', async () => {
      // Setup
      const mockFile = {
        buffer: Buffer.from(
          JSON.stringify({
            dependencies: {
              'vulnerable-package': '^1.0.0',
              'outdated-lib': '^2.0.0',
            },
          }),
        ),
      } as Express.Multer.File;

      // Execute
      const result = await service.analyzePublicRepository(
        'owner',
        'repo',
        mockFile,
      );

      // Assert
      expect(result.risky_dependencies).toContain('vulnerable-pkg@1.0.0');
      expect(result.dependency_health).toBe(75);
      expect(result.license_risks).toContain('GPL-3.0');
    });
  });

  describe('Data Persistence and Retrieval', () => {
    it('should maintain data consistency across operations', async () => {
      // First analysis
      const analysisResult = await service.analyzePublicRepository(
        'microsoft',
        'vscode',
      );

      // Retrieve via different methods
      const byId = await service.findOne('microsoft/vscode');
      const byOwner = await service.findByOwner('microsoft');
      const allRepos = await service.findAll();
      // Assert consistency
      expect(byId?.repo_id).toBe('microsoft/vscode');
      expect(byOwner?.[0]?.repo_id).toBe('microsoft/vscode');
      expect(allRepos?.data?.[0]?.repo_id).toBe('microsoft/vscode');
    });

    it('should provide accurate statistics', async () => {
      // Analyze multiple repositories
      await service.analyzePublicRepository('owner1', 'repo1');
      await service.analyzePublicRepository('owner2', 'repo2');
      await service.analyzePublicRepository('owner3', 'repo3');

      // Get stats
      const stats = await service.getStats();

      // Assert
      expect(stats.totalRepos).toBe(3);
      expect(stats.averageHealth).toBeGreaterThan(0);
      expect(stats.healthDistribution).toMatchObject({
        excellent: expect.any(Number),
        good: expect.any(Number),
        fair: expect.any(Number),
        poor: expect.any(Number),
      });
    });
  });

  describe('URL-based Analysis', () => {
    const testCases = [
      {
        url: 'https://github.com/facebook/react',
        expectedOwner: 'facebook',
        expectedRepo: 'react',
      },
      {
        url: 'https://github.com/microsoft/vscode.git',
        expectedOwner: 'microsoft',
        expectedRepo: 'vscode',
      },
      {
        url: 'git@github.com:nestjs/nest.git',
        expectedOwner: 'nestjs',
        expectedRepo: 'nest',
      },
    ];

    testCases.forEach(({ url, expectedOwner, expectedRepo }) => {
      it(`should correctly analyze repository from URL: ${url}`, async () => {
        const result = await service.analyzePublicRepoByUrl(url);

        expect(result.owner).toBe(expectedOwner);
        expect(result.repo).toBe(expectedRepo);
        expect(result.repo_id).toBe(`${expectedOwner}/${expectedRepo}`);
      });
    });
  });

  describe('Batch Processing', () => {
    it('should efficiently process multiple repositories with concurrency control', async () => {
      const requests = [
        { url: 'https://github.com/facebook/react' },
        { url: 'https://github.com/vuejs/vue' },
        { url: 'https://github.com/angular/angular' },
        { url: 'https://github.com/emberjs/ember.js' },
      ];

      const startTime = Date.now();
      const results = await service.analyzeMultipleRepositories(requests, 2);
      const duration = Date.now() - startTime;

      // Assert all were processed
      expect(results).toHaveLength(4);
      expect(results.every((r) => r.data !== undefined)).toBe(true);

      // Verify concurrency worked (should take less than sequential processing)
      expect(duration).toBeLessThan(6000); // 2 concurrent batches with delay
    }, 10000);
  });
});
