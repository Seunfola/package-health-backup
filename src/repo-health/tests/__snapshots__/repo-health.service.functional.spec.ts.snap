// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`RepoHealthService Functional should throw error if private repo is analyzed without token: Token is required for private repository 1`] = `"Token is required for private repository"`;
